pyiso/eu.py

```diff
@@ -0,0 +1,252 @@
+from pyiso.base import BaseClient
+import requests
+import pandas as pd
+from io import StringIO
+from time import sleep
+from datetime import datetime, timedelta
+import pytz
+from os import environ
+
+
+class EUClient(BaseClient):
+    NAME = 'EU'
+
+    base_url = '...'
+    TZ_NAME = 'UTC'
+
+    base_url = 'https://transparency.entsoe.eu/'
+    export_endpoint = 'load-domain/r2/totalLoadR2/export'
+
+    def handle_options(self, **kwargs):
+        # regular handle options
+        super(EUClient, self).handle_options(**kwargs)
+
+        # if latest is True
+        if self.options.get('latest', None):
+            self.options.update(start_at=datetime.now(pytz.utc) - timedelta(days=1),
+                                end_at=datetime.now(pytz.utc))
+
+        # workaround for base.handle_options setting forecast to false if end_at too far in past
+        if kwargs['forecast']:
+            self.options['forecast'] = True
+
+    def auth(self):
+        if not getattr(self, 'session', None):
+            self.session = requests.Session()
+
+        payload = {'j_username': environ['ENTSOe_USERNAME'],
+                   'j_password': environ['ENTSOe_PASSWORD']}
+
+        # Fake an ajax login to get the cookie
+        r = self.session.post(self.base_url + 'j_spring_security_check', params=payload,
+                              headers={'X-Ajax-call': 'true'})
+
+        msg = r.text
+        if msg == 'ok':
+            return True
+        elif msg == 'non_exists_user_or_bad_password':
+            # TODO throw error
+            return 'Wrong email or password'
+        elif msg == 'not_human':
+            return 'This account is not allowed to access web portal'
+        elif msg == 'suspended_use':
+            return 'User is suspended'
+        else:
+            return 'Unknown error:' + str(msg)
+
+    def fetch_entsoe(self, url, payload, count=0):
+        if not getattr(self, 'session', None):
+            self.auth()
+
+        r = self.session.get(url, params=payload)
+        # TODO error checking
+        if len(r.text) == 0:
+            if count > 3:  # try 3 times to get response
+                self.logger.warn('Request failed, no response found after %i attempts' % count)
+                return False
+            # throttled
+            sleep(5)
+            return self.fetch_entsoe(url, payload, count + 1)
+        if 'UNKNOWN_EXCEPTION' in r.text:
+            self.logger.warn('UNKNOWN EXCEPTION')
+            return False
+        return r.text
+
+    def construct_payload(self):
+        format_str = '%d.%m.%Y'
+        date_str = self.options['start_at'].strftime(format_str) + ' 00:00|UTC|DAY'
+        (TSO_ID, ) = [
+            i['ENTSOe_ID'] for i in control_areas if i['Code'] == self.options['control_area']]
```

Comment:
ETNSOe-ID's are two codes separated by a ! mark, but I can't figure out the pattern so I manually added every EU control area in a dict control_areas